sopel/modules/safety.py

```diff
@@ -212,80 +250,167 @@ def url_handler(bot: Sopel, trigger: Trigger):
+            elif not requested and r.status_code == 404:
+                # Not analyzed - submit new
+                LOGGER.debug("[VirusTotal] No scan for %r, requesting", safe_url)
+                r = requests.post(
+                    VT_API_URL,
+                    data={"url": url},
+                    headers={"x-apikey": bot.settings.safety.vt_api_key},
+                )
+                requested = True
+                sleep(2)  # Scans seem to take ~5s minimum, so add 2s
+        except requests.exceptions.RequestException:
+            # Ignoring exceptions with VT so domain list will always work
+            LOGGER.debug(
+                "[VirusTotal] Error obtaining response for %r", safe_url, exc_info=True
+            )
             return None
-        r.raise_for_status()
-        vt_data = r.json()
-    except requests.exceptions.RequestException:
-        # Ignoring exceptions with VT so domain list will always work
-        LOGGER.debug(
-            "[VirusTotal] Error obtaining response for %r", safe_url, exc_info=True
-        )
-    except json.JSONDecodeError:
-        # Ignoring exceptions with VT so domain list will always work
-        LOGGER.debug(
-            "[VirusTotal] Malformed response (invalid JSON) for %r",
-            safe_url,
-            exc_info=True,
-        )
-    fetched = time.time()
-    last_analysis = vt_data["data"]["attributes"]["last_analysis_stats"]
+        except json.JSONDecodeError:
+            # Ignoring exceptions with VT so domain list will always work
+            LOGGER.debug(
+                "[VirusTotal] Malformed response (invalid JSON) for %r",
+                safe_url,
+                exc_info=True,
+            )
+            return None
+        sleep(3)
+    else:  # Still no results
+        LOGGER.debug("[VirusTotal] Scan failed or unfinished for %r", safe_url)
+        return None
+    fetched = datetime.now(timezone.utc)
     # Only count strong opinions (ignore suspicious/timeout/undetected)
     result = {
         "positives": last_analysis["malicious"],
         "total": last_analysis["malicious"] + last_analysis["harmless"],
         "fetched": fetched,
-        "virustotal_data": vt_data,
+        # Subject to change formats!
+        "virustotal_data": vt_data["data"]["attributes"],
     }
     bot.memory["safety_cache"][url] = result
     if len(bot.memory["safety_cache"]) >= (2 * CACHE_LIMIT):
         _clean_cache(bot)
     return result
 
 
+@plugin.command("virustotal")
+@plugin.example(".virustotal https://malware.wicar.org/")
+@plugin.example(".virustotal hxxps://malware.wicar.org/")
+@plugin.output_prefix("[safety][VirusTotal] ")
+def vt_command(bot: Sopel, trigger: Trigger):
+    """Look up VT results on demand"""
+    if not bot.settings.safety.vt_api_key:
+        bot.reply("Sorry, I don't have a VirusTotal API key configured.")
+        return
+
+    url = trigger.group(2)
+    zwsp_safe_url = zwsp("hxx" + url[3:])
```

Comment:
I am very confused. Why replace the scheme prefix _and_ insert ZWSP between every character? Seems excessive, especially because `zwsp(thing)` as defined above will return `4 * len(thing) - 3` bytes when encoded to UTF-8 for sending to IRC (`'\u200b'.encode('utf-8')` ➡️ `b'\xe2\x80\x8b'`).

For nontrivial real-world URLs (not just the domain part) that might actually link to malware, this approach will quickly make the output of this command longer than the 510 total bytes available for one IRC line.

Reply:
The length concern is fair. The idea was that I don't know what clients use to decide what's a link or not (whether they'd catch just `example.com` out of `hxxp://example.com/`, and when quoting a possibly-malware domain we don't want it to ever be linkified.
I also didn't want to not refer to the link somehow when this may wait 10 sec for a scan & be confusing in a busy channel, though the same applies to the regular "WARNING:" lines.

```irc
<+mal> does everybody's client properly delinkify this? hxxp://example.com/
< RhinosF1> If I share it, it becomes a normal link
< RhinosF1> The protocol gets stripped
```

So, I think it's important to invalidate the link in at least one more way. I'll switch it to only ZWSP the periods.

Reply:
No more ZWSP, just hxxp and [.]ing the .s and :s